src/pudl/extract/extractor.py

```diff
@@ -0,0 +1,379 @@
+"""Generic functionality for extractors."""
+import importlib.resources
+from abc import abstractmethod
+from collections import defaultdict
+
+import pandas as pd
+from dagster import (
+    AssetsDefinition,
+    DynamicOut,
+    DynamicOutput,
+    OpDefinition,
+    graph_asset,
+    op,
+)
+
+import pudl
+
+StrInt = str | int
+PartitionSelection = list[StrInt] | tuple[StrInt] | StrInt
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+class GenericMetadata:
+    """Load generic metadata from Python package data.
+
+    When metadata object is instantiated, it is given ${dataset} name and it
+    will attempt to load csv files from pudl.package_data.${dataset} package.
+
+    It expects the following kinds of files:
+
+    * column_map/${page}.csv currently informs us how to translate input column
+      names to standardized pudl names for given (partition, input_col_name). Relevant
+      page is encoded in the filename.
+      
+    * column_map/${page}.csv currently informs us how to translate input column
+      names to standardized pudl names for given (partition, input_col_name).
+      Relevant page is encoded in the filename.
+    """
+
+    def __init__(self, dataset_name: str):
+        """Create Metadata object and load metadata from python package.
+
+        Args:
+            dataset_name: Name of the package/dataset to load the metadata from.
+            Files will be loaded from pudl.package_data.${dataset_name}
+        """
+        self._dataset_name = dataset_name
+        self._pkg = f"pudl.package_data.{dataset_name}"
+        column_map_pkg = self._pkg + ".column_maps"
+        self._column_map = {}
+        for res_path in importlib.resources.files(column_map_pkg).iterdir():
+            # res_path is expected to end with ${page}.csv
+            if res_path.suffix == ".csv":
+                column_map = self._load_csv(column_map_pkg, res_path.name)
+                self._column_map[res_path.stem] = column_map
+
+    def get_dataset_name(self) -> str:
+        """Returns the name of the dataset described by this metadata."""
+        return self._dataset_name
+
+    @staticmethod
+    def _load_csv(package: str, filename: str) -> pd.DataFrame:
+        """Load metadata from a filename that is found in a package."""
+        return pd.read_csv(
+            importlib.resources.files(package) / filename, index_col=0, comment="#"
+        )
+
+    @staticmethod
+    def _get_partition_selection(partition: dict[str, PartitionSelection]) -> str:
```

Comment:
`PartitionSelection` can potentially specify multiple attributes (e.g., `{'form': 'gas_distribution', 'year'='2020'}`)` but as far as I'm seeing, Excel and CSV extractors all expect a single attribute for getting the filename, validation, etc. I think if we need to we can override this in a subclass for more complex cases.